modules/core/src/smithy4s/http/HttpUnaryServerRouter.scala

```diff
@@ -22,91 +22,181 @@ import smithy4s.server.UnaryServerCodecs
+      getMethod,
+      getUri,
+      addDecodedPathParams
+    )
   }
 
-}
+  /**
+    * Creates a router in the form of a partial function, which is useful for integration with frameworks
+    * such as Play.
+    */
+  def partialFunction[Alg[_[_, _, _, _, _]], F[_], RequestHead, Request, Response](
+      service: smithy4s.Service[Alg]
+  )(
+      impl: service.Impl[F],
+      makeServerCodecs: UnaryServerCodecs.Make[F, Request, Response],
+      endpointMiddleware: Endpoint.Middleware[Request => F[Response]],
+      getMethod: RequestHead => HttpMethod,
+      getUri: RequestHead => HttpUri,
+      addDecodedPathParams: (Request, PathParams) => Request
+  )(implicit F: MonadThrowLike[F]): PartialFunction[RequestHead, Request => F[Response]] = {
+    new PartialFunctionRouter[Alg, service.Operation, F, RequestHead, Request, Response](
+      service,
+      service.toPolyFunction[smithy4s.kinds.Kind1[F]#toKind5](impl),
+      makeServerCodecs,
+      endpointMiddleware,
+      getMethod,
+      getUri,
+      addDecodedPathParams
+    )
+  }
 
-// scalafmt: {maxColumn = 120}
-private class HttpUnaryServerRouter[Alg[_[_, _, _, _, _]], Op[_, _, _, _, _], F[_], Request, Response](
-    service: smithy4s.Service.Aux[Alg, Op],
-    impl: FunctorInterpreter[Op, F],
-    makeServerCodecs: UnaryServerCodecs.Make[F, Request, Response],
-    endpointMiddleware: Endpoint.Middleware[Request => F[Response]],
-    getMethod: Request => HttpMethod,
-    getPathSegments: Request => IndexedSeq[String],
-    addDecodedPathParams: (Request, PathParams) => Request
-)(implicit F: MonadThrowLike[F]) {
-
-  @nowarn
-  private final case class HttpEndpointHandler(
-      httpEndpoint: HttpEndpoint[_],
-      handler: Request => F[Response]
-  )
-
-  def route(request: Request): Option[F[Response]] = {
-    val method = getMethod(request)
-    val pathSegments = getPathSegments(request)
-    perMethodEndpoint.get(method) match {
-      case Some(httpUnaryEndpoints) =>
-        val maybeMatched =
-          httpUnaryEndpoints.iterator
-            .map(ep => (ep.handler, ep.httpEndpoint.matches(pathSegments)))
-            .find(_._2.isDefined)
-        maybeMatched.flatMap {
-          case (handler, Some(pathParams)) =>
-            val amendedRequest = addDecodedPathParams(request, pathParams)
-            Some(handler(amendedRequest))
-          case (_, None) => None
-        }
+  private class KleisliRouter[Alg[_[_, _, _, _, _]], Op[_, _, _, _, _], F[_], Request, Response](
+      service: smithy4s.Service.Aux[Alg, Op],
+      impl: FunctorInterpreter[Op, F],
+      makeServerCodecs: UnaryServerCodecs.Make[F, Request, Response],
+      endpointMiddleware: Endpoint.Middleware[Request => F[Response]],
+      getMethod: Request => HttpMethod,
+      getUri: Request => HttpUri,
+      addDecodedPathParams: (Request, PathParams) => Request
+  )(implicit F: MonadThrowLike[F])
+      extends (Request => Option[F[Response]]) {
+
+    @nowarn
```

Comment:
**question:** what's the warning? Is it the `final` outer reference thing?

Perhaps worth a comment _or_ a more specific `@nowarn` rule (filter)